src/cpp/rtps/writer/StatelessWriter.cpp

```diff
@@ -1036,41 +593,176 @@ bool StatelessWriter::matched_reader_is_matched(
+                    group.sender(this, &*it);
+                    num_locators = it->locators_size();
+
+                    if (group.add_data(*cache_change, is_inline_qos_expected_))
+                    {
+                        add_statistics_sent_submessage(cache_change, num_locators);
+                    }
+                    else
+                    {
+                        logError(RTPS_WRITER, "Error sending change " << cache_change->sequenceNumber);
+                        ret_code = DeliveryRetCode::NOT_DELIVERED;
+                    }
+                }
+            }
+        }
+        else
+        {
+            if (0 < num_locators)
+            {
+                if (0 < n_fragments)
+                {
+                    for (FragmentNumber_t frag = current_fragment_sent_ + 1;
+                            DeliveryRetCode::DELIVERED == ret_code && frag <= n_fragments; ++frag)
+                    {
+                        if (group.add_data_frag(*cache_change, frag, is_inline_qos_expected_))
+                        {
+                            current_fragment_sent_ = frag;
+                            add_statistics_sent_submessage(cache_change, num_locators);
+                        }
+                        else
+                        {
+                            logError(RTPS_WRITER,
+                                    "Error sending fragment (" << cache_change->sequenceNumber << ", " << frag << ")");
+                            ret_code = DeliveryRetCode::NOT_DELIVERED;
+                        }
+                    }
+                }
+                else
+                {
+                    if (group.add_data(*cache_change, is_inline_qos_expected_))
+                    {
+                        add_statistics_sent_submessage(cache_change, num_locators);
+                    }
+                    else
+                    {
+                        logError(RTPS_WRITER, "Error sending change " << cache_change->sequenceNumber);
+                        ret_code = DeliveryRetCode::NOT_DELIVERED;
+                    }
+                }
+            }
+        }
+
+        on_sample_datas(cache_change->write_params.sample_identity(),
+                cache_change->writer_info.num_sent_submessages);
+        on_data_sent();
+
+    }
+    catch (const RTPSMessageGroup::timeout&)
+    {
+        logError(RTPS_WRITER, "Max blocking time reached");
+        ret_code = DeliveryRetCode::NOT_DELIVERED;
+    }
+    catch (const RTPSMessageGroup::limit_exceeded&)
+    {
+        ret_code = DeliveryRetCode::EXCEEDED_LIMIT;
+    }
+
+    group.sender(this, &locator_selector);
+
+    if (DeliveryRetCode::DELIVERED == ret_code &&
+            change_sequence_number > last_sequence_number_sent_)
+    {
+        if (nullptr != mp_listener)
+        {
+            mp_listener->onWriterChangeReceivedByAll(this, cache_change);
+        }
+
+        last_sequence_number_sent_ = change_sequence_number;
```

Comment:
We shuold notify `unsent_changes_cond_` here

Reply:
Done